File: packages/ld-adapter/src/__tests__/ld-api-client.test.ts
import { afterEach, describe, expect, it } from 'vitest';
import LaunchDarklyApi from 'launchdarkly-api';
import {
	createLaunchDarklyClient,
	DEFAULT_LD_API_VERSION,
} from '../client/ld-api-client.js';

describe('createLaunchDarklyClient', () => {
	afterEach(() => {
		const apiClient = LaunchDarklyApi.ApiClient.instance;
		delete apiClient.defaultHeaders['LD-API-Version'];
		delete apiClient.defaultHeaders['Content-Type'];
	});

	it('sets LD-API-Version header and respects baseUrl override', () => {
		createLaunchDarklyClient({
			accessToken: 'test-token',
			baseUrl: 'https://app.eu.launchdarkly.com',
		});

		const apiClient = LaunchDarklyApi.ApiClient.instance;
		expect(apiClient.defaultHeaders['LD-API-Version']).toBe(
			DEFAULT_LD_API_VERSION,
		);
		expect(apiClient.basePath).toBe('https://app.eu.launchdarkly.com');
		expect(apiClient.authentications.ApiKey.apiKey).toBe('test-token');
	});
});

File: kotlin/modules/ld-adapter/src/test/kotlin/com/ffpromo/ldadapter/client/LdApiClientTest.kt
package com.ffpromo.ldadapter.client

import com.ffpromo.contracts.LaunchDarklyClientConfig
import kotlinx.coroutines.test.runTest
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class LdApiClientTest {
    private lateinit var server: MockWebServer

    @BeforeEach
    fun setUp() {
        server = MockWebServer()
        server.start()
    }

    @AfterEach
    fun tearDown() {
        server.shutdown()
    }

    @Test
    fun `createLaunchDarklyClient sets LD-API-Version and Authorization on GET`() = runTest {
        server.enqueue(
            MockResponse()
                .setResponseCode(200)
                .setBody("""{"key":"sample-feature","variations":[],"environments":{}}"""),
        )

        val rawClient = createLaunchDarklyClient(
            LaunchDarklyClientConfig(
                accessToken = "test-token",
                baseUrl = server.url("/").toString().trimEnd('/'),
            ),
        )
        val client = LdApiClient(rawClient)
        client.getFeatureFlag("default", "sample-feature")

        val request = server.takeRequest()
        assertEquals("test-token", request.getHeader("Authorization"))
        assertEquals(DEFAULT_LD_API_VERSION, request.getHeader("LD-API-Version"))
        assertTrue(request.path!!.contains("/api/v2/flags/default/sample-feature"))
    }

    @Test
    fun `createLaunchDarklyClient respects baseUrl override`() {
        val rawClient = createLaunchDarklyClient(
            LaunchDarklyClientConfig(
                accessToken = "test-token",
                baseUrl = "https://app.eu.launchdarkly.com",
            ),
        )

        assertEquals("https://app.eu.launchdarkly.com", rawClient.config.baseUrl)
        assertEquals(DEFAULT_LD_API_VERSION, rawClient.config.apiVersion)
        assertEquals("test-token", rawClient.config.accessToken)
    }
}
